File: lib/trains-data/src/query.ts
import type { TrainEntry } from "./search";

export const MIN_QUERY_LENGTH = 2;

/** " 22943 " -> "22943"; "raj" -> "RAJ" */
export function normalizeTrainNumber(raw: string): string {
  return raw.replace(/\s+/g, "").toUpperCase().trim();
}

function normalizeName(value: string): string {
  return value.replace(/\s+/g, "").toLowerCase();
}

/** Exact number comparison on normalized values. */
export function matchesTrainNumber(
  suggestionNumber: string,
  normalizedInput: string,
): boolean {
  return normalizeTrainNumber(suggestionNumber) === normalizedInput;
}

/** Case/space-insensitive exact name comparison. */
export function matchesTrainName(
  suggestionName: string,
  normalizedInput: string,
): boolean {
  return normalizeName(suggestionName) === normalizeName(normalizedInput);
}

/** First occurrence wins per train number. */
export function uniqueByNumber(trains: TrainEntry[]): TrainEntry[] {
  const seen = new Set<string>();
  const unique: TrainEntry[] = [];
  for (const train of trains) {
    if (seen.has(train.number)) continue;
    seen.add(train.number);
    unique.push(train);
  }
  return unique;
}

/** Find a train by exact normalized number (dedupe-aware). */
export function findTrainByNumber(
  trains: TrainEntry[],
  number: string,
): TrainEntry | null {
  const normalized = normalizeTrainNumber(number);
  for (const train of trains) {
    if (normalizeTrainNumber(train.number) === normalized) return train;
  }
  return null;
}

File: rust/crates/trains-data/src/query.rs
//! Query normalization / matching helpers, ported 1:1 from `lib/trains-data/src/query.ts`.

use crate::search::TrainEntry;

/// Queries shorter than this are never searched. Ported from `MIN_QUERY_LENGTH`.
pub const MIN_QUERY_LENGTH: usize = 2;

/// `" 22943 "` -> `"22943"`; `"raj"` -> `"RAJ"`.
/// Strips all whitespace and uppercases, exactly like the TS `normalizeTrainNumber`.
pub fn normalize_train_number(raw: &str) -> String {
    raw.chars()
        .filter(|c| !c.is_whitespace())
        .collect::<String>()
        .to_uppercase()
}

fn normalize_name(value: &str) -> String {
    value
        .chars()
        .filter(|c| !c.is_whitespace())
        .collect::<String>()
        .to_lowercase()
}

/// Exact number comparison on normalized values.
pub fn matches_train_number(suggestion_number: &str, normalized_input: &str) -> bool {
    normalize_train_number(suggestion_number) == normalized_input
}

/// Case/space-insensitive exact name comparison.
pub fn matches_train_name(suggestion_name: &str, normalized_input: &str) -> bool {
    normalize_name(suggestion_name) == normalize_name(normalized_input)
}

/// First occurrence wins per train number. Returns owned clones in input order.
pub fn unique_by_number(trains: &[TrainEntry]) -> Vec<TrainEntry> {
    let mut seen = std::collections::HashSet::new();
    let mut unique: Vec<TrainEntry> = Vec::with_capacity(trains.len());
    for train in trains {
        if !seen.insert(&train.number) {
            continue;
        }
        unique.push(train.clone());
    }
    unique
}

/// Find a train by exact normalized number (dedupe-aware, first match wins).
pub fn find_train_by_number(trains: &[TrainEntry], number: &str) -> Option<&TrainEntry> {
    let normalized = normalize_train_number(number);
    trains
        .iter()
        .find(|train| normalize_train_number(&train.number) == normalized)
}
